File: netlify/functions/upload_csv.py
import json
import pandas as pd
import io
import base64

def is_valid_number(val):
    try:
        float(val)
        return True
    except:
        return False

def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
        csv_b64 = body.get("file")
        if not csv_b64:
            return {"statusCode": 400, "body": json.dumps({"error": "Missing file"})}

        csv_data = base64.b64decode(csv_b64).decode("utf-8", errors="ignore")
        df = pd.read_csv(io.StringIO(csv_data))

        df = df.drop(columns=["Company"], errors="ignore")
        expected = [
            "Ticker",
            "%Ch Curr Qtr Est - 4 wks",
            "%Ch Next Qtr  Est - 4 wks",
            "%Ch Curr Fiscal Yr Est - 4 wks",
            "# Up Curr Fiscal Yr Est - 4 wks"
        ]
        df = df[[c for c in expected if c in df.columns]]

        cleaned = []
        for _, row in df.iterrows():
            ticker = str(row.get("Ticker", "")).strip()
            if not ticker or ticker.lower() == "nan" or ticker.lower() == "average":
                continue

            vals = [
                row.get("%Ch Curr Qtr Est - 4 wks"),
                row.get("%Ch Next Qtr  Est - 4 wks"),
                row.get("%Ch Curr Fiscal Yr Est - 4 wks"),
                row.get("# Up Curr Fiscal Yr Est - 4 wks")
            ]

            if not all(is_valid_number(v) for v in vals):
                continue

            cleaned.append({
                "ticker": ticker,
                "pct_change_current_quarter": float(vals[0]),
                "pct_change_next_quarter": float(vals[1]),
                "pct_change_current_fiscal_year": float(vals[2]),
                "num_current_fiscal_year_estimates": int(float(vals[3]))
            })

        return {"statusCode": 200, "body": json.dumps({"preview": cleaned})}
    except Exception as e:
        return {"statusCode": 500, "body": json.dumps({"error": str(e)})}

File: netlify/functions/upload_csv.mjs
import { parse } from "csv-parse/sync";

const EXPECTED_COLUMNS = [
  "Ticker",
  "%Ch Curr Qtr Est - 4 wks",
  "%Ch Next Qtr  Est - 4 wks",
  "%Ch Curr Fiscal Yr Est - 4 wks",
  "# Up Curr Fiscal Yr Est - 4 wks",
];

const isValidNumber = (val) => {
  if (val === null || val === undefined || val === "") return false;
  const n = Number(String(val).trim());
  return Number.isFinite(n);
};

export default async (req) => {
  try {
    const body = await req.json().catch(() => ({}));
    const fileB64 = body.file;
    if (!fileB64) {
      return Response.json({ error: "Missing file" }, { status: 400 });
    }

    const csvText = Buffer.from(fileB64, "base64").toString("utf-8");

    const records = parse(csvText, {
      columns: true,
      skip_empty_lines: true,
      relax_column_count: true,
      trim: true,
    });

    const cleaned = [];
    for (const row of records) {
      const ticker = String(row.Ticker ?? "").trim();
      if (!ticker || ticker.toLowerCase() === "nan" || ticker.toLowerCase() === "average") {
        continue;
      }

      const vals = EXPECTED_COLUMNS.slice(1).map((col) => row[col]);
      if (!vals.every(isValidNumber)) continue;

      cleaned.push({
        ticker,
        pct_change_current_quarter: Number(vals[0]),
        pct_change_next_quarter: Number(vals[1]),
        pct_change_current_fiscal_year: Number(vals[2]),
        num_current_fiscal_year_estimates: Math.trunc(Number(vals[3])),
      });
    }

    return Response.json({ preview: cleaned });
  } catch (e) {
    return Response.json({ error: e.message }, { status: 500 });
  }
};

export const config = {
  path: "/.netlify/functions/upload_csv",
};
